src/main/java/org/rumbledb/compiler/DynamicContextVisitor.java

```diff
@@ -71,8 +72,8 @@ public DynamicContext visit(Node node, DynamicContext argument) {
     @Override
     public DynamicContext visitFunctionDeclaration(FunctionDeclaration declaration, DynamicContext argument) {
         InlineFunctionExpression expression = (InlineFunctionExpression) declaration.getExpression();
-        Map<String, SequenceType> paramNameToSequenceTypes = new LinkedHashMap<>();
-        for (Map.Entry<String, SequenceType> paramEntry : expression.getParams().entrySet()) {
+        Map<Name, SequenceType> paramNameToSequenceTypes = new LinkedHashMap<>();
+        for (Map.Entry<Name, SequenceType> paramEntry : expression.getParams().entrySet()) {
```

Comment:
I want to ask a question with the hopes of learning from you on the topic of OOP & readability:

I found myself wondering about the potential benefits and drawbacks of subclassing the 'Name' class. I believe creating FunctionName, ParamName, and VariableName subclasses could benefit us with improved readability and type safety. (eg. we could simply call the LinkedHashMap (paramToSequenceTypes since type information would already indicate that the keys are FunctionNames). But I was wondering if you can point out some drawbacks with this extended hierarchy such as performance or maintenance.

Reply:
That's an excellent point and good thinking.

There is a benefit of subclassing in terms of code quality. There is a benefit in readability of the consuming code.

There is a drawback in performance as class hierarchies are slower than flat classes.  There is drawback in readability of the producing code as developers need to open several files instead of just one.

In this case, the code in the Name class is executed billions of time, as it is massively used while accessing the dynamic context -> this is why performance considerations win.

In a case where the code is executed rarely (such as client-side or desktop application), the benefits would start to outweigh the drawbacks.

If you are interested in seeing the difference of performance, I have a sandbox here:

https://github.com/ghislainfourny/PerformanceMeasurements